extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,439 @@
+      this.selectItem_(event.target);
+    });
+  }
+
+  /**
+   * Render filtered results on the current input and update the container_.
+   * @private
+   */
+  renderResults_() {
+    const userInput = this.inputElement_.value;
+    this.clearAllItems_();
+    if (userInput.length < this.minChars_ || !this.inlineData_) {
+      return;
+    }
+    const filteredData = this.filterData_(this.inlineData_, userInput);
+    filteredData.forEach(item => {
+      this.container_.appendChild(this.createElementFromItem_(item));
+    });
+
+    // Append the partial user-provided input to navigate back to.
+    this.container_.appendChild(this.createElementFromItem_(userInput));
+  }
+
+  /**
+   * Apply the filter to the given data based on the given input.
+   * @param {!Array<string>} data
+   * @param {string} input
+   * @return {!Array<string>}
+   * @private
+   */
+  filterData_(data, input) {
+    input = input.toLowerCase();
+    let filteredData = data.filter(item => {
+      switch (this.filter_) {
+        case FilterType.SUBSTRING:
+          return includes(item, input);
+        case FilterType.PREFIX:
+          return startsWith(item, input);
+        case FilterType.TOKEN_PREFIX:
+          return item.split(' ').some(token => {
+            return startsWith(token, input);
+          });
+        case FilterType.FUZZY:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.CUSTOM:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.NONE:
+          // Query server endpoint.
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        default:
+          throw new Error(`Unexpected filter: ${this.filter_}`);
+      }
+    });
+
+    // Truncate to max-entries.
+    if (this.maxEntries_ && this.maxEntries_ < filteredData.length) {
+      filteredData = filteredData.slice(0, this.maxEntries_);
+    }
+
+    return filteredData;
+  }
+
+  /**
+   * Shows or hides the results container_.
+   * @param {boolean=} opt_display
+   * @private
+   */
+  toggleResults_(opt_display) {
+    if (!this.container_) {
+      return;
+    }
+    toggle(this.container_, opt_display);
+  }
+
+  /**
+   * Handle showing or hiding results on user focus/blur.
+   * @param {boolean} opt_display
+   * @return {!Promise}
+   * @private
+   */
+  toggleResultsHandler_(opt_display) {
+    return this.mutateElement(() => {
+      if (!opt_display) {
+        this.resetActiveElement_();
+        this.activeIndex_ = -1;
+      }
+      this.toggleResults_(opt_display);
+    });
+  }
+
+  /**
+   * Returns true if the results are visible and has items.
+   * @return {boolean}
+   * @private
+   */
+  resultsShowing_() {
+    return !this.container_.hasAttribute('hidden') &&
+      this.container_.children.length > 0;
+  }
+
+  /**
+   * Returns true if the given element is a suggested item.
+   * @param {Element|EventTarget} element
```

Comment:
nit: object types like `Element` or `EventTarget` should be explicitly non-nullable `!Element`  or nullable `?Element`, but not implicitly nullable `{Element}`. As annotated, this method states that it accepts `null` as a value  for the `element` parameter, but that would cause a reference error.